### PR TITLE
Credits Roll

### PR DESCRIPTION
**Pull request by Cyberboss**
:cl:
add: Added the credits roll
/:cl:

Preview: https://www.youtube.com/watch?v=HUS-jPKCAxo&feature=youtu.be

TODO:
- [x] Potentially see if this can be overlayed directly on client huds (they are currently objs fixed to the starting splash sceen) so they can be played when the round ends
- [x] Figure out how to make it work with Identicons
- [x] Figure out why some images (my own for example) throw an array exception: Depends on #30012 
- [x] Set up automatic processes for getting github contributors list, downloading images, and creating .dmi

### CODE REVIEW

**Comment by The-clown-known-as-Pepe** [comment]
cheater

**Comment by ecronogist** [comment]
Excellent numbers, friendo.

**Comment by InsaneHyena** [comment]
30000 get

**Comment by PraiseRatvar** [comment]
dreammaker_2017-08-17_23-50-45
Thanks to
@Adrinus 
@anconfuzedrock 
@apuh3 
@Arianya 
@astralenigma 
@A-t48 
@catalystfd
@Ccomp5950 
@Cheridan 
@ChuckTheSheep 
@CosmicScientist 
@ExcessiveUseOfCobblestone 
@havocx42 
@Iamgoofball 
@Impisi 
@johnsmith38 
@KazeEspada 
@kevinz000 
@Lexorion 
@Limeliz 
@majorsephiroth 
@melichior 
@Mloc 
@MMMiracles 
@MoreRobustThanYou 
@N3X15 
@Neerti 
@Nienhaus 
@Niknakflak 
@nivomi 
@Nodrak 
@norill 
@nullquery 
@ojacobson 
@Okand37 
@optimumtact 
@ORCACommander 
@p440 
@Perakp 
@pubby 
@RemieRichards 
@Rumia29 
@sawu-tg 
@Sieve 
@spudboy555 
@Summoners 
@SuperSayu 
@SweaterKittens 
@Tastyfish 
@tgstation-server (bots!?)

|  |  | 
VVV

**Comment by PraiseRatvar** [comment]
@theOperand 
@TheVekter 
@Thunder12345 
@tikisux 
@Tofa01 
@Ultimate-Chimera 
@Uristqwerty 
@WhiteHusky 
@WitheredGryphon 
@Xhuis 
@YotaXP 
@Zelacks 
@Zerrien 
and... @Zuvassin 
for your involvement within the /tg/station coding 'community', we hope you enjoyed working with everyone on the repository! 🎉 🍰 🎉

**Comment by Cyberboss** [comment]
This isn't even half the people, I need to refine the process

**Comment by MoreRobustThanYou** [comment]
and hopefully will continue!

**Comment by DoctoreMad** [comment]
Wait why are you giving thanks to a robot

**Comment by PraiseRatvar** [comment]
@DoctoreMad its basically mso too

**Comment by DoctoreMad** [comment]
that works

File: code/game/objects/effects/credits.dm [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#define CREDIT_ROLL_SPEED 150
+#define CREDIT_SPAWN_SPEED 30
+#define CREDIT_ANIMATE_HEIGHT (14 * TURF_PIXEL_DIAMETER)
+#define CREDIT_EASE_DURATION 30
+#define CREDITS_LOC locate(12, 167, ZLEVEL_CENTCOM)
+
+/proc/RollCredits()
+    var/icon/credits_file = new('icons/credits.dmi')
+    var/list/contributors = icon_states(credits_file) - null
+    qdel(credits_file)
```
Comment (MrPerson) [inline code/game/objects/effects/credits.dm]: This could have just been var/list/contributors = icon_states('icons/credits.dmi') - null

**Comment by A-t48** [comment]
:o

**Comment by kilkun** [comment]
can i be on the list

**Comment by JJRcop** [comment]
😕

**Comment by optimumtact** [comment]
@PraiseRatvar can you not

**Comment by Qustinnus** [comment]
nice quads my dude

**Comment by Cyberboss** [comment]
OP updated with preview vid

**Comment by TheVekter** [comment]
This is a good PR.

Please change TheVekter to just Vekter tia

**Comment by optimumtact** [comment]
yeah we should allow a config mapping between github name -> the name I want which is oranges

**Comment by Cyberboss** [comment]
Updated OP with piratepad link for name remaps

**Comment by Durkel** [comment]
Nice get and with a somewhat non meme pr.

**Comment by Ultimate-Chimera** [comment]
Huh

**Comment by CosmicScientist** [comment]
A mention? I'm relevant?!

*Ahem*
That is pretty sweeet.

**Comment by Cyberboss** [comment]
@PraiseRatvar the necromancer

**Comment by lordpidey** [comment]
I don't seem to be on that list?  Is there a way to join it?

**Comment by FuryMcFlurry** [comment]
What about spriters who cant code?

**Comment by pubby** [comment]
> Figure out why some images (my own for example) throw an array exception here

Your avatar has no alpha channel, so maybe you're interpreting RGB data as RGBA.

File: code/__DEFINES/misc.dm [hunk 2]
```diff
@@ -441,3 +441,5 @@ GLOBAL_LIST_INIT(ghost_others_options, list(GHOST_OTHERS_SIMPLE, GHOST_OTHERS_DE
 #define MOUSE_OPACITY_TRANSPARENT 0
 #define MOUSE_OPACITY_ICON 1
 #define MOUSE_OPACITY_OPAQUE 2
+
+#define TURF_PIXEL_DIAMETER 32
```
Comment (RemieRichards) [inline code/__DEFINES/misc.dm]: this is just ```world.icon_size```, no?

File: code/game/objects/effects/credits.dm [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+/obj/effect/abstract/credit/Initialize(mapload, credited)
+	. = ..()
+	name = text("\improper []", credited)
+	icon_state = credited
+	maptext = credited
+	animate(src, pixel_y = CREDIT_ANIMATE_HEIGHT, time = CREDIT_ROLL_SPEED)
+	animate(src, alpha = 255, time = CREDIT_EASE_DURATION, flags = ANIMATION_PARALLEL)
+	addtimer(CALLBACK(src, .proc/FadeOut), CREDIT_ROLL_SPEED - CREDIT_EASE_DURATION)
+	QDEL_IN(src, CREDIT_ROLL_SPEED)
```
Comment (RemieRichards) [inline code/game/objects/effects/credits.dm]: ```QDEL_IN_CLIENT_TIME``` ?

**Comment by XDTM** [comment]
Put me in the screenshot

**Comment by Cyberboss** [comment]
I'll post the whole thing when I get in

**Comment by Cyberboss** [comment]
@alexkar598 
```
java -jar dmitool.jar import credits.dmi Cyberboss pngs/Cyberboss.png
Loading credits.dmi
1 images, 1 states, size 32x32
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 96
        at dmitool.IconState.importFromPNG(IconState.java:200)
        at dmitool.Main.main(Main.java:374)
```
Here's the file I'm using: https://www.dropbox.com/s/a689vmyerv16flf/Cyberboss.png?dl=0

File: code/game/objects/effects/credits.dm (hunk 3 above)
Comment (Cyberboss) [inline code/game/objects/effects/credits.dm]: I tried using client time timers for the whole thing, but it ended up being really unreliable as opposed to regular timers. @MrStonedOne ?

**Comment by alexkar598** [comment]
@Cyberboss

Il try to debug the program when I get to an actual computer.

**Comment by alexkar598** [comment]
Il try to debug the program when I get to a computer in about 3-4 days if you haven't fixed it then

**Comment by Supermichael777** [comment]
no player references

**Comment by MMMiracles** [comment]
it's OK, coders don't play

**Comment by kilkun** [comment]
then why do we have a ton of george melons and cuban pete references

**Comment by Cyberboss** [comment]
@pubby that's probably exactly what's happening, I'll see if I can fix up dmitool

**Comment by TheVekter** [comment]
Can you fucking mongoloids take the ancient argument to Discord or
something thanks

**Comment by NewSta1** [comment]
What are the requirements for being there?

**Comment by MoreRobustThanYou** [comment]
contributing i think

**Comment by NewSta1** [comment]
I contributed two PRs does that count

**Comment by Cyberboss** [comment]
If you have a commit linked to your github account in this repo, you will show up

**Comment by TheVekter** [comment]
Presumably, having something committed to the game.

On Aug 18, 2017 1:46 PM, "MoreRobustThanYou" <notifications@github.com>
wrote:

contributing i think

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/30000#issuecomment-323417761>,
or mute the thread

**Comment by theOperand** [comment]
Nice, my name will actually be on something for once :^]

Pls change my name to Miauw, I've added it in the PiratePad.

**Comment by NewSta1** [comment]
Welp, my two PRs were actually from forks I made to tg - would that count?

**Comment by Cyberboss** [comment]
If you have [Contributor] next to your name, you'll be in

**Comment by Cyberboss** [comment]
untitled

Now the main problem with the identicons is they aren't friendly to the ?s=32 parameter in the URL. So we'll have to resize them ourselves somehow

**Comment by Cyberboss** [comment]
e.g. https://avatars2.githubusercontent.com/u/4176358?v=4&s=32
vs https://avatars2.githubusercontent.com/u/8171642?v=4&s=32

**Comment by RemieRichards** [comment]
report it as a bug, hack around it for now.

**Comment by norill** [comment]
apparently there are no PMs on github, so i'll just post here: @theOperand get back to KAG miauw

**Comment by Cyberboss** [comment]
untitled

NOW we're talking!

**Comment by NewSta1** [comment]
@Cyberboss We have a bit of a problem here.

As I understand, the contributors list accounts only for those who DIRECTLY committed to tgstation, instead of trying to merge their forked branches.

As I fall into the latter category, I am not in the list of contributors to tgstation, and I suspect I am not the only one.

Any way to fix this?

**Comment by Supermichael777** [comment]
Only lists the top 100 dont worry

**Comment by Cyberboss** [comment]
@Supermichael777 no this method lists all of them

@NewSta1 You are in the credits

**Comment by optimumtact** [comment]
@Cyberboss call me Oranges one more time see what happens

**Comment by optimumtact** [comment]
cheeky fuckers

**Comment by Cyberboss** [comment]
Context: someone fucked with oranges entry on piratepad and capitalized the o

File: tools/CreditsTool/remappings.txt [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+# e.g.
+# Cyberboss __REMOVE__
+
+optimumtact Oranges
+qustinnus Floyd / Qustinnus
+catalystfd __REMOVE__
+TheVekter Vekter
+tgstation-server __REMOVE__
+ChangelingRain Joan
+kebab __REMOVE__
```
Comment (JJRcop) [inline tools/CreditsTool/remappings.txt]: 🎺 🎺 🎺 🎹

**Comment by nicbn** [comment]
Could you use python instead of .net? Linux users exist

**Comment by TheVekter** [comment]
Linux users are a myth perpetrated by Mac users to make Windows users
uncomfortable.

On Aug 18, 2017 11:42 PM, "nicbn" <notifications@github.com> wrote:

Could you use python instead of .net? Linux users exist

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/30000#issuecomment-323497431>,
or mute the thread

**Comment by nicbn** [comment]
Also make the names closer together, two hours of scrolling letters is not nice.

**Comment by Armhulen** [comment]
so, can this be triggered to roll with a verb? it looks kinda shitty if it goes there every single time.

**Comment by nicbn** [comment]
This could be shown at round end.

File: tools/CreditsTool/remappings.txt (hunk 4 above)
Comment (RemieRichards) [inline tools/CreditsTool/remappings.txt]: this was me, I'm glad you like it.
(when I saw the comment about __REMOVE__ I couldn't resist)

**Comment by JJRcop** [comment]
@nicbn Isn't .net on linux nowadays? https://github.com/dotnet/core

**Comment by Cyberboss** [comment]
If mono can't run this simple-ass program it has no hope

**Comment by Arianya** [comment]
@Cyberboss I can provide a 32x32 of my icon or the full original image as needed.

File: tools/CreditsTool/remappings.txt (hunk 4 above)
Comment (Cyberboss) [inline tools/CreditsTool/remappings.txt]: Um, does this need to be changed? I'm out of the loop

Comment (RemieRichards) [inline tools/CreditsTool/remappings.txt]: kebab isn't a person, this is a joke entry I put in (tbh I'm REALLY surprised you missed it)

Comment (Cyberboss) [inline tools/CreditsTool/remappings.txt]: https://github.com/kebab wew

**Comment by Cyberboss** [comment]
You're fine as you are @Arianya

File: tools/CreditsTool/remappings.txt (hunk 4 above)
Comment (Iamgoofball) [inline tools/CreditsTool/remappings.txt]: @Cyberboss https://www.youtube.com/watch?v=SxUU3zncVmI

**Comment by PraiseRatvar** [comment]
https://soundcloud.com/admiralhippie/space-station-13
tbh this needs to play when credits roll
tbh some kind of song needs to play when credits roll

**Comment by Cyberboss** [comment]
@FrozenGuy5 #30029

On Aug 21, 2017 12:01 PM, "FrozenGuy5" <notifications@github.com> wrote:

https://soundcloud.com/admiralhippie/space-station-13
tbh this needs to play when credits roll
tbh some kind of song needs to play when credits roll

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/30000#issuecomment-323799191>,
or mute the thread

**Comment by calzilla1** [comment]
Will this be a pref?

**Comment by optimumtact** [comment]
I don't think we should give people a choice, like how we force show everyone the changelog

**Comment by kilkun** [comment]
It better include all of the other devs who contributed stuff, stretching back 14 years. I really want to read those names.

**Comment by kilkun** [comment]
And also at the very end it should say "...And you!! Thanks for playing!" with a big collage of beloved ss13 characters

**Comment by calzilla1** [comment]
Well, what if an admin is doing memes in the back ground?

**Comment by kilkun** [comment]
delay round end.

**Comment by Shadowlight213** [comment]
>not in the credits.
Ok, you're not getting the AI lawset storage refactor anymore.

**Comment by Shadowlight213** [comment]
Also, why have this at the end of every round? Why not make the contributors list in the changelog auto update?

**Comment by Steelpoint** [comment]
This is nice. Gives credit to the long list of people who've contributed to the game.

**Comment by JJRcop** [comment]
@Shadowlight213 You're in.
I put a lot of effort in making this look like a screenshot, but it's stitched together :D

**Comment by PraiseRatvar** [comment]
added my name to piratepad, also who the fuck capitalized oranges' name.

**Comment by Cyberboss** [comment]
There is a verb to cancel it once it starts

**Comment by Cyberboss** [comment]
@kilkun There's too many people to show the entire list everytime. That's why it's randomized

**Comment by MrStonedOne** [comment]
Getting a list of everybody who's contributed from github isn't easy, the contributor api is broken, it should be showing up to 500 results, but it shows 0 to 23 depending on its mood.

A list of everybody who's submitted a pr would be easier.

**Comment by Cyberboss** [comment]
I've been pulling 318 consistently since I started this PR, which seems about right. Been using this link 
https://api.github.com/repos/tgstation/tgstation/contributors?page=1 which paginates to 11

**Comment by Shadowlight213** [comment]
\>github api has a response var just to specify if the user is staff.
I feel like it could have been handled better.

**Comment by MrStonedOne** [comment]
cyberboss: https://api.github.com/repos/tgstation/tgstation/contributors?per_page=100

**Comment by nicbn** [comment]
Remember to iterate the pages with `&page=n`

**Comment by Bawhoppen** [comment]
Does this have functionality to selectively purge people we don't like anymore? (no names named)

**Comment by optimumtact** [comment]
why on earth would it do that

**Comment by Bawhoppen** [comment]
Why wouldn't it

**Comment by Cyberboss** [comment]
Don't rewrite history. It has that function but it's for a different purpose

**Comment by lzimann** [comment]
why a verb to cancel instead of making this a button to play(like it was suggested, a button in the lobby screen)?

**Comment by RemieRichards** [comment]
forcing it to play makes it like a movie/film which is good imo.

**Comment by lzimann** [comment]
it's annoying, not good

**Comment by RemieRichards** [comment]
that's your opinion.

**Comment by lzimann** [comment]
this is delaying the round restart until credits finish

**Comment by Cyberboss** [comment]
@lzimann it's in OOC

**Pull request merged**